### PR TITLE
protocols/{relay,dcutr}: Replace `std::time::SystemTime` with `instant::SystemTime`

### LINKED ISSUE #2984: Error running a relay v2 client on wasm32-unknown-unknown
## Summary

Running a relay v2 client on `wasm32-unknown-unknown` gives a panic:
```
panicked at 'time not implemented on this platform', library/std/src/sys/wasm/../unsupported/time.rs:31:9
```

## Expected behaviour

Relay client works on `wasm32-unknown-unknown`

## Actual behaviour

Running a test with wasm-pack gives the following output, looks like the problem comes from https://github.com/libp2p/rust-libp2p/blob/master/protocols/relay/src/v2/protocol/outbound_hop.rs#L136

<details><summary>Debug Output</summary>
<p>

```
running 1 test

test wasm_net::test::browser1_node1_browser2_1::browser1 ... FAIL

failures:

---- wasm_net::test::browser1_node1_browser2_1::browser1 output ----
    log output:
        %cDEBUG%c src/test/common.rs:111%c node1_peer_id: 12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c src/test/common.rs:115%c local_peer_id: 12D3KooWP285Hw3CSTdr9oU6Ezz4hDoi6XS5vfDjjNeTJ1uFMGvp color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:134%c Dialer: Proposed protocol: /noise color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:183%c Dialer: Received confirmation for protocol: /noise color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/core/src/upgrade/apply.rs:231%c Successfully applied negotiated protocol color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:134%c Dialer: Proposed protocol: /yamux/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:183%c Dialer: Received confirmation for protocol: /yamux/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/.cargo/registry/src/github.com-1ecc6299db9ec823/yamux-0.10.2/src/connection.rs:277%c new connection: 2467ee42 (Client) color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/core/src/upgrade/apply.rs:231%c Successfully applied negotiated protocol color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/swarm/src/lib.rs:694%c Connection established: PeerId("12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN") Dialer { address: "/ip4/127.0.0.1/tcp/38615/p2p/12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN/ws", role_override: Dialer }; Total (peer): 1. Total non-banned (peer): 1 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/.cargo/registry/src/github.com-1ecc6299db9ec823/yamux-0.10.2/src/connection.rs:536%c 2467ee42: new outbound (Stream 2467ee42/1) of (Connection 2467ee42 Client (streams 0)) color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:134%c Dialer: Proposed protocol: /ipfs/ping/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/listener_select.rs:231%c Listener: confirming protocol: /ipfs/ping/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/listener_select.rs:293%c Listener: sent confirmed protocol: /ipfs/ping/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/core/src/upgrade/apply.rs:151%c Successfully applied negotiated protocol color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:183%c Dialer: Received confirmation for protocol: /ipfs/ping/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/core/src/upgrade/apply.rs:231%c Successfully applied negotiated protocol color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/protocols/ping/src/lib.rs:136%c Ping received from PeerId("12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN") color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/protocols/ping/src/lib.rs:135%c Ping sent to PeerId("12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN") color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/listener_select.rs:231%c Listener: confirming protocol: /ipfs/id/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/listener_select.rs:293%c Listener: sent confirmed protocol: /ipfs/id/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/core/src/upgrade/apply.rs:151%c Successfully applied negotiated protocol color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c src/test/common.rs:162%c Told relay its public address. color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/.cargo/registry/src/github.com-1ecc6299db9ec823/yamux-0.10.2/src/connection.rs:536%c 2467ee42: new outbound (Stream 2467ee42/3) of (Connection 2467ee42 Client (streams 2)) color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:134%c Dialer: Proposed protocol: /ipfs/id/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:183%c Dialer: Received confirmation for protocol: /ipfs/id/1.0.0 color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/core/src/upgrade/apply.rs:231%c Successfully applied negotiated protocol color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c src/test/common.rs:169%c Relay told us our public address: "/ip4/127.0.0.1/tcp/44724/ws" color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/.cargo/registry/src/github.com-1ecc6299db9ec823/yamux-0.10.2/src/connection.rs:536%c 2467ee42: new outbound (Stream 2467ee42/5) of (Connection 2467ee42 Client (streams 3)) color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:134%c Dialer: Proposed protocol: /libp2p/circuit/relay/0.2.0/hop color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
        %cDEBUG%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/misc/multistream-select/src/dialer_select.rs:183%c Dialer: Received confirmation for protocol: /libp2p/circuit/relay/0.2.0/hop color: white; padding: 0 3px; background: blue; font-weight: bold; color: inherit background: inherit; color: inherit
    
    warn output:
        %cWARN%c /home/fgimenez/workspace/nillion/libp2p-poc/libp2p/swarm/src/lib.rs:1405%c Failed to create executor thread pool: Error { kind: Unsupported, message: "operation not supported on this platform" } color: white; padding: 0 3px; background: orange; font-weight: bold; color: inherit background: inherit; color: inherit
    
    error output:
        panicked at 'time not implemented on this platform', library/std/src/sys/wasm/../unsupported/time.rs:31:9
        
        Stack:
        
        Error
            at http://localhost:8000/wasm-bindgen-test:640:21
            at logError (http://localhost:8000/wasm-bindgen-test:221:18)
            at imports.wbg.__wbg_new_abda76e883ba8a5f (http://localhost:8000/wasm-bindgen-test:639:66)
            at console_error_panic_hook::Error::new::h3a63fa2ff713d487 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[30923]:0xab1519)
            at console_error_panic_hook::hook_impl::hf240382615ecac40 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[4314]:0x6fbce9)
            at console_error_panic_hook::hook::h67f84998deaf2b99 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[33823]:0xad93ec)
            at core::ops::function::Fn::call::h51d34082a6c99473 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[29153]:0xa958c8)
            at std::panicking::rust_panic_with_hook::hb09154fa23e06c37 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[10149]:0x88cd83)
            at std::panicking::begin_panic_handler::{{closure}}::h6091c197f0d08bf0 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[13965]:0x92da4a)
            at std::sys_common::backtrace::__rust_end_short_backtrace::h004afb3e6a867c40 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[39531]:0xb10d9d)
        
        
    
    JS exception that was thrown:
        RuntimeError: unreachable
            at __rust_start_panic (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[39999]:0xb1249b)
            at rust_panic (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[36797]:0xafc0d7)
            at std::panicking::rust_panic_with_hook::hb09154fa23e06c37 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[10149]:0x88cdae)
            at std::panicking::begin_panic_handler::{{closure}}::h6091c197f0d08bf0 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[13965]:0x92da4a)
            at std::sys_common::backtrace::__rust_end_short_backtrace::h004afb3e6a867c40 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[39531]:0xb10d9d)
            at rust_begin_unwind (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[32491]:0xac7bb8)
            at core::panicking::panic_fmt::h9e229748e3ae9f9d (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[32932]:0xacdbae)
            at std::time::SystemTime::now::hfae2007f187ae317 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[32930]:0xacdb2b)
            at <libp2p_relay::v2::protocol::outbound_hop::Upgrade as libp2p_core::upgrade::OutboundUpgrade<multistream_select::negotiated::Negotiated<libp2p_core::muxing::boxed::SubstreamBox>>>::upgrade_outbound::{{closure}}::he20a566948833906 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[131]:0x1ec466)
            at <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::he739fe8783e77660 (http://localhost:8000/wasm-bindgen-test_bg.wasm:wasm-function[7308]:0x7ea60d)

failures:

    wasm_net::test::browser1_node1_browser2_1::browser1

test result: FAILED. 0 passed; 1 failed; 0 ignored
```
</p>
</details>

## Version

- libp2p version (version number, commit, or branch): 0.49.0, da0403dc455536a4593f88da0052e883be91573a

## Would you like to work on fixing this bug?

Yes!

### PR DESCRIPTION
**Pull request by thomaseizinger**
# Description

I tried to safeguard against this by banning `std::time::SystemTime::now` via clippy.toml but unfortunately, it doesn't work because `instant` re-exports `std` when not compiling for wasm so it still gets flagged.

## Links to any relevant issues

- Fixes https://github.com/libp2p/rust-libp2p/issues/2984

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/relay/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,10 @@
 
 - Update to `libp2p-swarm` `v0.40.0`.
 
+- Fix WASM compilation. See [PR XXXX].
+
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX/
```
Comment (thomaseizinger) [inline protocols/relay/CHANGELOG.md]:
```suggestion
- Fix WASM compilation. See [PR 2991].

[PR 2991]: https://github.com/libp2p/rust-libp2p/pull/2991/
```

**Comment by fgimenez** [comment]
Thanks for taking care @thomaseizinger I'm getting panics bc of this one too https://github.com/libp2p/rust-libp2p/blob/master/protocols/dcutr/src/protocol/outbound.rs#L29

File: protocols/dcutr/src/protocol/outbound.rs [hunk 2]
```diff
@@ -22,11 +22,11 @@ use crate::message_proto::{hole_punch, HolePunch};
 use asynchronous_codec::Framed;
 use futures::{future::BoxFuture, prelude::*};
 use futures_timer::Delay;
+use instant::Instant;
```
Comment (mxinden) [inline protocols/dcutr/src/protocol/outbound.rs:25]: Missing changelog entry.

**Review by mxinden — APPROVED** [review]
Just one small nit.

Thanks @thomaseizinger for the fix.

@fgimenez I am not sure how your WASM setup looks like. In case your local WASM node can not listen for incoming connections, I don't see a way how you can succeed in hole punching and thus I don't see the need for `libp2p-dcutr`. I am not saying we shouldn't do this change. Just a heads up. This will e.g. change for the browser world once we have WebRTC support. See https://github.com/libp2p/specs/pull/412.

**Comment by fgimenez** [comment]
> @fgimenez I am not sure how your WASM setup looks like. 

Thanks for the heads up @mxinden, let me explain a bit the setup I've come up with, please don't hesitate to let me know if anything is wrong or should be done in a different way. My goal was to verify connectivity between WASM browser nodes through a relay v2 node, I was aiming to achieve something similar to what was described in this comment https://github.com/libp2p/js-libp2p-webrtc-direct/issues/98#issuecomment-759494593. I have a wasm-pack based test case that spins up 2 browser nodes and a backend relay node, the setup looks like this:
* Each browser node runs as a separate wasm-pack process with headless chrome. The code they run is based on https://github.com/libp2p/rust-libp2p/blob/master/protocols/dcutr/examples/dcutr.rs and both nodes are set in listen mode (they are instructed to listen on the relayed multiaddr) with two differences:
  * the transport uses `libp2p_wasm_ext::ExtTransport::new(libp2p_wasm_ext::ffi::websocket_transport())`
  * clients are not set to listen on all interfaces, only on the relayed multiaddr.
* The backend relay server is based on https://github.com/libp2p/rust-libp2p/blob/master/protocols/relay/examples/relay_v2.rs
* All the nodes run locally

Regarding the test case itself:
* One of the browsers dials on the relayed multiaddr of the other, something like: `/ip4/127.0.0.1/tcp/38615/ws/p2p/12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN/p2p-circuit/p2p/12D3KooWP285Hw3CSTdr9oU6Ezz4hDoi6XS5vfDjjNeTJ1uFMGvp`
* each browser node checks that it receives a ping from the other.

The test is passing with the changes in this PR, I've attached logs for the 3 nodes
[node-logs.zip](https://github.com/libp2p/rust-libp2p/files/9744677/node-logs.zip)

> In case your local WASM node can not listen for incoming connections, I don't see a way how you can succeed in hole punching and thus I don't see the need for libp2p-dcutr. I am not saying we shouldn't do this change. Just a heads up. This will e.g. change for the browser world once we have WebRTC support. See https://github.com/libp2p/specs/pull/412.

As explained above, this setup is working fine just with websockets, and as I understand it, it requires libp2p-dcutr to work, is that correct? please let me know if I'm missing something, or if this is not the right approach.

**Comment by mxinden** [comment]
> As explained above, this setup is working fine just with websockets, and as I understand it, it requires libp2p-dcutr to work, is that correct? please let me know if I'm missing something, or if this is not the right approach.

@fgimenez I am not sure I am following. Is the final connection between the two nodes a relayed or a direct connection?

DCUtR upgrades a relayed connection to a direct connection. In case you can not establish a direct connection (between two browsers), I don't see how DCUtR can help you.

https://github.com/libp2p/specs/pull/412 will give you browser-to-server connectivity. For your WASM node you would need to adapt `ExtTransport` for the browser's WebRTC stack. This is a medium size undertaking. Contributions would be greatly appreciated.

Long term, https://github.com/libp2p/specs/pull/412 will give you browser-to-browser connectivity including hole punching.

**Comment by fgimenez** [comment]
> @fgimenez I am not sure I am following. Is the final connection between the two nodes a relayed or a direct connection?
> 
> DCUtR upgrades a relayed connection to a direct connection. In case you can not establish a direct connection (between two browsers), I don't see how DCUtR can help you.

@mxinden you are right, one of the browsers is dialing the other on the relayed address, following the hole-punching example I added the DCUtR behaviour but I can see in the logs how the upgrade fails, the error comes from https://github.com/libp2p/rust-libp2p/blob/master/protocols/dcutr/src/protocol/inbound.rs#L57
I was confused because the test just checks for ping events, and each browser is able to receive a ping from the other one, will check more deeply how this works.

> [libp2p/specs#412](https://github.com/libp2p/specs/pull/412) will give you browser-to-server connectivity. For your WASM node you would need to adapt `ExtTransport` for the browser's WebRTC stack. This is a medium size undertaking. Contributions would be greatly appreciated.

Awesome, would love to work on it, should be similar to what is done here for websockets right? https://github.com/libp2p/rust-libp2p/tree/master/transports/wasm-ext/src 

> Long term, [libp2p/specs#412](https://github.com/libp2p/specs/pull/412) will give you browser-to-browser connectivity including hole punching.

Very cool :+1:

**Comment by mxinden** [comment]
> > [libp2p/specs#412](https://github.com/libp2p/specs/pull/412) will give you browser-to-server connectivity. For your WASM node you would need to adapt `ExtTransport` for the browser's WebRTC stack. This is a medium size undertaking. Contributions would be greatly appreciated.
> 
> Awesome, would love to work on it, should be similar to what is done here for websockets right? https://github.com/libp2p/rust-libp2p/tree/master/transports/wasm-ext/src

Yes. Correct. Though it does include an additional Noise handshake and message framing. See https://github.com/libp2p/specs/pull/412.

**Pull request merged**